respy/python/solve/solve_auxiliary.py

```diff
@@ -568,44 +677,152 @@ def check_input(respy_obj):
+    type_deviations = types_dummy.dot(optim_paras["type_shifts"][:, :2])
+
+    states.wage_a = states.wage_a * np.exp(type_deviations.iloc[:, 0])
+    states.wage_b = states.wage_b * np.exp(type_deviations.iloc[:, 1])
+
+    return states
+
+
+class StateSpace:
+    """This class represents a state space.
+
+    Attributes
+    ----------
+    states : pd.DataFrame
+    indexer : np.array
+
+    """
+
+    def create_state_space(self, *args):
+        """This function creates the state space.
+
+        It should return two objects. First, a DataFrame where each row represents one
+        state. This object captures characteristics of states in a dense format. The
+        second object is a Numpy matrix where each dimension represents one
+        characteristic of the state and the values contain the indices of the state in
+        the DataFrame.
 
-    return wages
+        """
+        self.states, self.indexer = pyth_create_state_space(*args)
+
+    def create_covariates(self):
+        self.states = create_covariates(self.states)
+
+    @property
+    def states_per_period(self):
+        return self.states.groupby("period").count().iloc[:, 0].values
+
+    def _get_fortran_counterparts(self):
+        periods_rewards_systematic = np.full(
+            (self.states_per_period.shape[0], self.states_per_period.max(), 4),
+            np.nan,
+        )
+        for period, group in self.states.groupby("period"):
+            sub = group[
+                [
+                    "rewards_systematic_a",
+                    "rewards_systematic_b",
+                    "rewards_systematic_edu",
+                    "rewards_systematic_home",
+                ]
+            ].values
+
+            periods_rewards_systematic[period, : sub.shape[0]] = sub
+
+        periods_emax = np.full(
+            (self.states_per_period.shape[0], self.states_per_period.max(), 4),
+            np.nan,
+        )
+        for period, group in self.states.groupby("period"):
+            sub = group[
+                ["emaxs_a", "emaxs_b", "emaxs_edu", "emaxs_home"]
+            ].values
+
+            periods_emax[period, : sub.shape[0], :] = sub
+
+        mapping_state_idx = self.indexer
+        states_all = None
+
+        return (
+            periods_rewards_systematic,
+            self.states_per_period,
+            mapping_state_idx,
+            periods_emax,
+            states_all,
+        )
+
+    def __len__(self):
+        return len(self.states)
+
+    def __getitem__(self, key):
+        """
+
+        Raises
+        ------
+        InadmissibleStateError
+            The error is raised if a non-valid indexer is used. The simplest way to
+            raise this error is to assign -1 to every inadmissible state. This value is
+            still an integer to keep memory costs down.
+
+        """
+        key = tuple(int(i) for i in key)
+        position = self.indexer[key]
+
+        try:
+            return self.states.loc[position]
```

Comment:
I compared the lookup speed with .loc, .iloc and if the states were a numpy array. There is almost no difference between .loc and .iloc, but the array lookup is about 5 times faster. I don't really understand why, but in the long run we should look for a way to make the lookup faster.

Reply:
If ``.iloc`` is faster than ``.loc``, I will change it. I thought there is a reason to use ``.loc``, but there is none.

Reply:
There is a reason. .loc[-1] will raise a KeyError, .iloc[-1] probably won't. So you might have to handle that case explicitly.

Reply:
Quick performance check:
1. loc
2. iloc
3. iloc with if check whether position is equal to -1.
4. states is numpy array

![image](https://user-images.githubusercontent.com/22533006/53087821-598e0180-3508-11e9-9eeb-aedc95f6b220.png)

For now, we can use the if statement and get a slight boost. But given that each state except in the last period performs a lookup for four states, we have - assuming a state space with 40 periods, one type and education starting at ten to 20 - 300.000 states * 4 * 131e-6 seconds = 157 seconds. Maybe this is a new bottleneck and we have to drop the dataframe again.

Reply:
Before we think about dropping it, we can still try to optimize the lookup. For example, we could store everything in numpy arrays internally, but the lookup could still return a pd.Series. Maybe we don't even have to create the Series in each lookup but could just overwrite the data of a Series we initialize once.

Reply:
![image](https://user-images.githubusercontent.com/22533006/53092325-bf808600-3514-11e9-9251-93a4525fc87c.png)

If this comparison is valid, we would not get any performance boost. The construction of the series takes to much time.

Reply:
![image](https://user-images.githubusercontent.com/9536380/53093337-6c5c0280-3517-11e9-9b99-915d8353d4a9.png)

Reply:
This will be fleshed out in a separate PR.